### PR TITLE
Use Redis SIGSTOP supervision to avoid sleep() call

### PR DESCRIPTION
**Pull request by mehrdadn**
## Why are these changes needed?

Redis [already has built-in functionality](https://github.com/antirez/redis/blob/27e5f385c1839157574b80f2079d79bf40e32639/redis.conf#L137-L146) for to allow a supervising process (like `upstart`) to await its startup.  This mechanism helps make the code timing more deterministic and consistent, and removes unnecessary delays incurred by `sleep()`.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/services.py [hunk 1]
```diff
@@ -945,17 +945,30 @@ def _start_redis_instance(executable,
...
             stderr_file=stderr_file,
             fate_share=fate_share)
-        time.sleep(0.1)
-        # Check if Redis successfully started (or at least if it the executable
-        # did not exit within 0.1 seconds).
+        if wait_for_sigstop:
+            # Redis signals when it's ready by pausing itself; wait for it
+            status = os.waitpid(process_info.process.pid, os.WUNTRACED)[1]
+            if os.WIFEXITED(status):
+                break
```
Comment (simon-mo) [inline python/ray/services.py:966]: this should fail to `counter == num_retries`

File: python/ray/services.py [hunk 2]
```diff
@@ -945,17 +945,30 @@ def _start_redis_instance(executable,
...
             command += ["--requirepass", password]
-        command += (
-            ["--port", str(port), "--loglevel", "warning"] + load_module_args)
+        command += ["--port", str(port), "--loglevel", "warning"]
+        command += ["--supervised", "upstart"]
+        command += load_module_args
+        wait_for_sigstop = sys.platform != "win32"
+        env_updates = {}
+        if wait_for_sigstop:
+            env_updates.update(UPSTART_JOB="")
```
Comment (simon-mo) [inline python/ray/services.py:954]: can we link to some documentation about this in redis repo?

Comment (mehrdadn) [reply python/ray/services.py:955]: Done

File: python/ray/services.py (hunk 1 above)
Comment (mehrdadn) [reply python/ray/services.py]: I think I might've actually had a bug here (I didn't seem to account for port allocation failure?), but I pushed a fix that should hopefully be clearer, see my update.

**Comment by mehrdadn** [comment]
Closing because this mechanism is fundamentally broken: it seems that Redis lies to the supervisor about its readiness. It [sends the notification _before_ it initializes itself](https://github.com/redis/redis/blob/5.0.9/src/server.c#L4196-L4201):
```
    server.supervised = redisIsSupervised(server.supervised_mode);
    int background = server.daemonize && !server.supervised;
    if (background) daemonize();

    initServer();
    if (background || server.pidfile) createPidFile();
```
This means when e.g. port binding fails, the supervisor has already been falsely notified that Redis has successfully initialized, so this mechanism doesn't let us remove the timeout.

**Pull request closed without merging**